{
  "type": "issue",
  "issue": {
    "id": 1928881368,
    "node_id": "I_kwDOABII585y-GTY",
    "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/28603",
    "repository_url": "https://api.github.com/repos/bitcoin/bitcoin",
    "labels_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/28603/labels%7B/name%7D",
    "comments_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/28603/comments",
    "events_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/28603/events",
    "html_url": "https://github.com/bitcoin/bitcoin/issues/28603",
    "number": 28603,
    "state": "open",
    "state_reason": null,
    "title": "Extend feature_init.py file perturbations",
    "body": "### Motivation\r\n\r\n`feature_init.py` is a functional test that perturbs various files in the data directory (simulation data corruption), and checks the response during startup.\r\n\r\nRecently, this test was improved in #27823, but there were some open suggestions in, in particular https://github.com/bitcoin/bitcoin/pull/27823#discussion_r1267289758, to extend and/or randomize the way files are perturbed.\r\n\r\nThere is also the possibility that more sophisticated perturbations could potentially uncover other unhandled exceptions/assert crashes like the one fixed in #27823.  \r\n\r\n### Possible solution\r\n\r\n_No response_\r\n\r\n### Useful Skills\r\n\r\n* Python\r\n* Debugging the bitcoin core C++ file parsing code in order to understand why certain perturbation lead to certain errors.\r\n\r\n\r\n### Guidance for new contributors\r\n\r\nWant to work on this issue?\r\n\r\nFor guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.\r\n",
    "user": {
      "login": "mzumsande",
      "id": 48763452,
      "node_id": "MDQ6VXNlcjQ4NzYzNDUy",
      "avatar_url": "https://avatars.githubusercontent.com/u/48763452?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/mzumsande",
      "html_url": "https://github.com/mzumsande",
      "followers_url": "https://api.github.com/users/mzumsande/followers",
      "following_url": "https://api.github.com/users/mzumsande/following%7B/other_user%7D",
      "gists_url": "https://api.github.com/users/mzumsande/gists%7B/gist_id%7D",
      "starred_url": "https://api.github.com/users/mzumsande/starred%7B/owner%7D%7B/repo%7D",
      "subscriptions_url": "https://api.github.com/users/mzumsande/subscriptions",
      "organizations_url": "https://api.github.com/users/mzumsande/orgs",
      "repos_url": "https://api.github.com/users/mzumsande/repos",
      "events_url": "https://api.github.com/users/mzumsande/events%7B/privacy%7D",
      "received_events_url": "https://api.github.com/users/mzumsande/received_events",
      "type": "User",
      "site_admin": false
    },
    "labels": [
      {
        "id": 349312277,
        "node_id": "MDU6TGFiZWwzNDkzMTIyNzc=",
        "url": "https://api.github.com/repos/bitcoin/bitcoin/labels/good%20first%20issue",
        "name": "good first issue",
        "color": "c2e0c6",
        "default": true
      }
    ],
    "assignees": [],
    "author_association": "CONTRIBUTOR",
    "locked": false,
    "comments": 0,
    "created_at": "2023-10-05T19:00:50Z",
    "updated_at": "2023-10-05T19:01:05Z"
  },
  "events": [
    {
      "event": "labeled",
      "id": 10567472989,
      "node_id": "LE_lADOABII585y-GTYzwAAAAJ13tdd",
      "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/events/10567472989",
      "actor": {
        "login": "mzumsande",
        "id": 48763452,
        "node_id": "MDQ6VXNlcjQ4NzYzNDUy",
        "avatar_url": "https://avatars.githubusercontent.com/u/48763452?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/mzumsande",
        "html_url": "https://github.com/mzumsande",
        "followers_url": "https://api.github.com/users/mzumsande/followers",
        "following_url": "https://api.github.com/users/mzumsande/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/mzumsande/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/mzumsande/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/mzumsande/subscriptions",
        "organizations_url": "https://api.github.com/users/mzumsande/orgs",
        "repos_url": "https://api.github.com/users/mzumsande/repos",
        "events_url": "https://api.github.com/users/mzumsande/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/mzumsande/received_events",
        "type": "User",
        "site_admin": false
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2023-10-05T19:00:50Z",
      "label": {
        "name": "good first issue",
        "color": "c2e0c6"
      }
    },
    {
      "event": "cross-referenced",
      "id": null,
      "node_id": null,
      "url": null,
      "actor": {
        "login": "mzumsande",
        "id": 48763452,
        "node_id": "MDQ6VXNlcjQ4NzYzNDUy",
        "avatar_url": "https://avatars.githubusercontent.com/u/48763452?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/mzumsande",
        "html_url": "https://github.com/mzumsande",
        "followers_url": "https://api.github.com/users/mzumsande/followers",
        "following_url": "https://api.github.com/users/mzumsande/following%7B/other_user%7D",
        "gists_url": "https://api.github.com/users/mzumsande/gists%7B/gist_id%7D",
        "starred_url": "https://api.github.com/users/mzumsande/starred%7B/owner%7D%7B/repo%7D",
        "subscriptions_url": "https://api.github.com/users/mzumsande/subscriptions",
        "organizations_url": "https://api.github.com/users/mzumsande/orgs",
        "repos_url": "https://api.github.com/users/mzumsande/repos",
        "events_url": "https://api.github.com/users/mzumsande/events%7B/privacy%7D",
        "received_events_url": "https://api.github.com/users/mzumsande/received_events",
        "type": "User",
        "site_admin": false
      },
      "commit_id": null,
      "commit_url": null,
      "created_at": "2023-10-05T19:01:31Z",
      "updated_at": "2023-10-05T19:01:31Z",
      "source": {
        "issue": {
          "id": 1740531376,
          "node_id": "PR_kwDOABII585SIOMw",
          "url": "https://api.github.com/repos/bitcoin/bitcoin/issues/27823",
          "repository_url": "https://api.github.com/repos/bitcoin/bitcoin",
          "labels_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/27823/labels%7B/name%7D",
          "comments_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/27823/comments",
          "events_url": "https://api.github.com/repos/bitcoin/bitcoin/issues/27823/events",
          "html_url": "https://github.com/bitcoin/bitcoin/pull/27823",
          "number": 27823,
          "state": "closed",
          "state_reason": null,
          "title": "init: return error when block index is non-contiguous, fix feature_init.py file perturbation",
          "body": "When the block index database is non-contiguous due to file corruption (i.e. it contains indexes of height `x-1` and `x+1`, but not `x`), bitcoind can currently crash with an assert in `BuildSkip()` / `GetAncestor()` during `BlockManager::LoadBlockIndex()`:\r\n```\r\nbitcoind: chain.cpp:112: const CBlockIndex* CBlockIndex::GetAncestor(int) const: Assertion `pindexWalk->pprev' failed.\r\n```\r\nThis PR changes it such that we instead return an `InitError` to the user.\r\n\r\nI stumbled upon this because I noticed that the file perturbation in `feature_init.py`  wasn't working as intended, which is fixed in the second commit:\r\n* Opening the file twice in one `with` statement would lead to `tf_read` being empty, so the test wouldn't perturb anything but replace the file with a new one. Fixed by first opening for read, then for write.\r\n* We need to restore the previous state after perturbations, so that only the current perturbation is active and not a mix of the current and previous ones.\r\n* I also added `checkblocks=200` to the startup parameters so that corruption in earlier blocks of `blk00000.dat` is detected during init verification and not ignored.\r\n\r\nAfter fixing `feature_init.py` like that I'd run into the `assert` mentioned above (so running the testfix from the second commit without the first one is a way to reproduce it).",
          "user": {
            "login": "mzumsande",
            "id": 48763452,
            "node_id": "MDQ6VXNlcjQ4NzYzNDUy",
            "avatar_url": "https://avatars.githubusercontent.com/u/48763452?v=4",
            "gravatar_id": "",
            "url": "https://api.github.com/users/mzumsande",
            "html_url": "https://github.com/mzumsande",
            "followers_url": "https://api.github.com/users/mzumsande/followers",
            "following_url": "https://api.github.com/users/mzumsande/following%7B/other_user%7D",
            "gists_url": "https://api.github.com/users/mzumsande/gists%7B/gist_id%7D",
            "starred_url": "https://api.github.com/users/mzumsande/starred%7B/owner%7D%7B/repo%7D",
            "subscriptions_url": "https://api.github.com/users/mzumsande/subscriptions",
            "organizations_url": "https://api.github.com/users/mzumsande/orgs",
            "repos_url": "https://api.github.com/users/mzumsande/repos",
            "events_url": "https://api.github.com/users/mzumsande/events%7B/privacy%7D",
            "received_events_url": "https://api.github.com/users/mzumsande/received_events",
            "type": "User",
            "site_admin": false
          },
          "labels": [],
          "assignees": [],
          "author_association": "CONTRIBUTOR",
          "locked": false,
          "comments": 8,
          "pull_request": {
            "url": "https://api.github.com/repos/bitcoin/bitcoin/pulls/27823",
            "html_url": "https://github.com/bitcoin/bitcoin/pull/27823",
            "diff_url": "https://github.com/bitcoin/bitcoin/pull/27823.diff",
            "patch_url": "https://github.com/bitcoin/bitcoin/pull/27823.patch"
          },
          "closed_at": "2023-10-04T19:37:12Z",
          "created_at": "2023-06-04T18:55:13Z",
          "updated_at": "2023-10-05T19:01:30Z"
        },
        "type": "issue"
      }
    }
  ]
}